File: algorithms/2_medium/C/01641_count_sorted_vowel_strings/dp.c
int countVowelStrings(const int n) {
    if (n < 1 || n > 50) { return -1; }

    int buff[10] = { 0 };
    int *restrict prev = &buff[0];
    for (int i = 0; i < 5; i += 1) { prev[i] = 1; }
    int *restrict curr = &buff[5];

    for (int steps = 1; steps < n; steps += 1) {
        for (int i = 0; i < 5; i += 1) { 
            curr[i] = 0;
        }

        for (int i = 0; i < 5; i += 1) {
            for (int j = 0; j <= i; j += 1) {
                curr[j] += prev[i];
            }
        }

        void *temp = prev;
        prev = curr;
        curr = temp;
    }

    int sum = 0;
    for (int i = 0; i < 5; i += 1) {
        sum += prev[i];
    }

    return sum;
}

File: algorithms/2_medium/C/01641_count_sorted_vowel_strings/dp.rs
impl Solution {
    pub fn count_vowel_strings(n: i32) -> i32 {
        assert!(n >= 1 && n <= 50);
        let mut prev = [1_i32; 5];
        let mut curr = [0_i32; 5];

        for _ in 1..n {
            curr.iter_mut().for_each(|x| *x = 0);
            for i in 0..5 {
                for j in 0..=i {
                    curr[j] += prev[i];
                }
            }
            (curr, prev) = (prev, curr);
        }

        prev.into_iter().sum()
    }
}
